repository/src/lib.rs

```diff
@@ -156,13 +156,184 @@ impl<T: RawRepository> DistributedRepository<T> {
+        let last_header = self.get_last_finalized_block_header().await?;
+        let reserved_state = self.get_reserved_state().await?;
+
+        let mut block_header_vec = vec![];
+        let mut finalized_branches = vec![];
+
+        'branch: for branch in branches {
+            let branch_tip_commit_hash = self.raw.locate_branch(branch.clone()).await?;
+            // Check if the branch is rebased on top of the `finalized` branch.
+            if finalized_branch_commit_hash
+                != self
+                    .raw
+                    .find_merge_base(branch_tip_commit_hash, finalized_branch_commit_hash)
+                    .await?
+            {
+                self.raw.delete_branch(branch.to_string()).await?;
+                continue;
+            }
+
+            // Construct a commit list starting from the next commit of the last finalized block to the `branch_tip_commit`(the most recent commit of the branch)
+            let ancestor_commits = self
+                .raw
+                .list_ancestors(branch_tip_commit_hash, Some(256))
+                .await?;
+            let position = ancestor_commits
+                .iter()
+                .position(|c| *c == finalized_branch_commit_hash)
+                .expect("TODO: handle the case where it exceeds the limit");
+            let new_semantic_commits = stream::iter(
+                ancestor_commits
+                    .iter()
+                    .take(position)
+                    .rev()
+                    .cloned()
+                    .map(|c| {
+                        let raw = &self.raw;
+                        async move { raw.read_semantic_commit(c).await.map(|x| (x, c)) }
+                    }),
+            )
+            .buffered(256)
+            .collect::<Vec<_>>()
+            .await;
+            let mut new_semantic_commits = new_semantic_commits
+                .into_iter()
+                .collect::<Result<Vec<_>, _>>()?;
+            // Add most recent commit of the branch to the list since it is not included in the ancestor commits
+            let branch_tip_semantic_commit = self
+                .raw
+                .read_semantic_commit(branch_tip_commit_hash)
+                .await?;
+            new_semantic_commits.push((branch_tip_semantic_commit, branch_tip_commit_hash));
+            let new_commits = new_semantic_commits
+                .into_iter()
+                .map(|(commit, hash)| {
+                    from_semantic_commit(commit)
+                        .map_err(|e| (e, hash))
+                        .map(|x| (x, hash))
+                })
+                .collect::<Result<Vec<_>, _>>()
+                .map_err(|(error, hash)| {
+                    anyhow!("failed to convert the commit {}: {}", hash, error)
+                })?;
+
+            // Verify all the incoming commits and apply them to the local repository only if they are valid.
```

Comment:
I'm not sure but probably a logic to not verify if it is agenda/block branch needs to be added

Reply:
You mean we should add comment? or it is about the code?

Reply:
a code for such logic

Reply:
I think we just have to verify agenda/block branch too. How do you think about it?